sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java

```diff
@@ -272,52 +338,315 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
+      /**
+       * It adds a record asynchronously which then should be delivered by Kinesis producer in
+       * background (Kinesis producer forks native processes to do this job).
+       *
+       * <p>The records can be batched and then they will be sent in one HTTP request. Amazon KPL
+       * supports two types of batching - aggregation and collection - and they can be configured by
+       * producer properties.
+       *
+       * <p>More details can be found here:
+       * <a href="https://docs.aws.amazon.com/streams/latest/dev/kinesis-kpl-concepts.html">KPL Key Concepts</a> and
+       * <a href="https://docs.aws.amazon.com/streams/latest/dev/kinesis-kpl-config.html">Configuring the KPL</a>
+       */
+      @ProcessElement
+      public void processElement(ProcessContext c) throws Exception {
+        checkForFailures();
+
+        // Need to avoid keeping too many futures in producer's map to prevent OOM.
+        // In usual case, it should exit immediately.
+        flush(MAX_NUM_RECORDS);
+
+        ByteBuffer data = ByteBuffer.wrap(c.element());
+        String partitionKey = spec.getPartitionKey();
+        String explicitHashKey = null;
+
+        // Use custom partitioner
+        if (partitioner != null) {
+          partitionKey = partitioner.getPartitionKey(c.element());
+          explicitHashKey = partitioner.getExplicitHashKey(c.element());
         }
-        return clientBuilder.build();
+
+        ListenableFuture<UserRecordResult> f =
+            producer.addUserRecord(spec.getStreamName(), partitionKey, explicitHashKey, data);
+        Futures.addCallback(f, new UserRecordResultFutureCallback());
       }
 
-      @Override
-      public AmazonCloudWatch getCloudWatchClient() {
-        AmazonCloudWatchClientBuilder clientBuilder =
-            AmazonCloudWatchClientBuilder.standard().withCredentials(getCredentialsProvider());
-        if (serviceEndpoint == null) {
-          clientBuilder.withRegion(region);
-        } else {
-          clientBuilder.withEndpointConfiguration(
-              new AwsClientBuilder.EndpointConfiguration(serviceEndpoint, region.getName()));
+      @FinishBundle
+      public void finishBundle() throws Exception {
+        checkForFailures();
+
+        // Flush all outstanding records, blocking call
+        flushAll();
+      }
+
+      @Teardown
+      public void tearDown() throws Exception {
+        if (producer != null) {
+          producer.destroy();
+          producer = null;
+        }
+      }
+
+      /**
+       * Flush outstanding records until the total number will be less than required or
+       * number of retries will be exhausted.
+       */
+      private void flush(int numMax) throws InterruptedException, IOException {
+        int retries = spec.getRetries();
+        int numOutstandingRecords = producer.getOutstandingRecordsCount();
+
+        while (numOutstandingRecords > numMax && retries-- > 0) {
```

Comment:
In what case would `producer.flush()` not flush everything - why do we even need a retry loop? I think new records cannot be added while this runs because they're only added from `ProcessElement`.

Reply:
Let me explain how actually `KinesisProducer` flushes records. When we create new instance of `KinesisProducer`, it spawns new child process (this is a native process) which, in fact, does all work to aggregate and to send the records to Kinesis. In its order, `KinesisProducer` "talks" to this process through pipes. 

So, once we call `producer.flush()` then it just sends a command to child process to flush outstanding records but with no guarantee that the buffered records will be sent. For this, we need to run a loop to check the number of outstanding records to make sure that all or required number records were sent.

Actually, `KinesisProducer` has another method - `producer.flushSync()` - which just runs infinite loop until all messages will be flushed:
```java
    public void flushSync() {
        while (getOutstandingRecordsCount() > 0) {
            flush();
            try {
                Thread.sleep(500);
            } catch (InterruptedException e) { }
        }
    }
```
but I'm not sure it's a good idea to use this method to flush everything in `TearDown` or `FinishBundle` since it might block forever. That is why I created a new method which tries to flush only the number of retry times and it will throw an exception in case if it still has **more than required** number of outstanding records and I use it to flush all bundle records in `FinishBundle` (`flushAll()` equal to `flush(0)`).

Also, I call this method `flush(MAX_NUM_RECORDS)` in `ProcessElement` to prevent unlimited growing the buffer of added futures that are stored in `HashMap` (as we discussed it before). The goal of this call is to wait until number of outgoing messages becomes less than `MAX_NUM_RECORDS` and yes, until this, we can't add new records. In the first version of PR, I didn't have this call but, potentially, we could have OOM there, so it was added.

Reply:
Oh I see, I didn't realize flush was asynchronous. In that case the current code is fine, modulo removal of the tuning knob (I think it should be sufficient to use a default of e.g. retrying for up to 1 minute - I can't think of a case where it would be critically important to use a different value; I'm happy to reconsider if such a case emerges in practice).

Reply:
Ok, I changed a total time of retrying to 1 minute. I believe it should quite enough for most cases as a default value.